src/zasm.zig

```diff
@@ -676,6 +676,180 @@ pub const Pcid = packed struct {
     }
 };
 
+// ~~~ [ zig-x86_64/src/structures/paging/page_table.zig ] ~~~
+
+/// A 64-bit page table entry.
+pub const PageTableEntry = packed struct {
+    entry: u64,
+
+    /// Creates an unused page table entry.
+    pub fn init() PageTableEntry {
+        return .{ .entry = 0 };
+    }
+
+    /// Returns whether this entry is zero.
+    pub fn isUnused(self: PageTableEntry) bool {
+        return self.entry == 0;
+    }
+
+    /// Sets this entry to zero.
+    pub fn setUnused(self: *PageTableEntry) void {
+        self.entry = 0;
+    }
+
+    /// Returns the flags of this entry.
+    pub fn getFlags(self: PageTableEntry) PageTableFlags {
+        return PageTableFlags.fromU64(self.entry);
+    }
+
+    /// Returns the physical address mapped by this entry, might be zero.
+    pub fn getAddr(self: PageTableEntry) PhysAddr {
+        // Unchecked is used as the mask ensures validity
+        return PhysAddr.initUnchecked(self.entry & 0x000f_ffff_ffff_f000);
+    }
+
+    /// Returns the physical frame mapped by this entry.
+    ///
+    /// Returns the following errors:
+    ///
+    /// - `FrameError::FrameNotPresent` if the entry doesn't have the `present` flag set.
+    /// - `FrameError::HugeFrame` if the entry has the `huge_page` flag set (for huge pages the
+    ///    `addr` function must be used)
+    pub fn getFrame(self: PageTableEntry) FrameError!PhysFrame {
+        const flags = self.getFlags();
+
+        if (!flags.present) {
+            return FrameError.FrameNotPresent;
+        }
+
+        if (flags.huge) {
+            return FrameError.HugeFrame;
+        }
+
+        return PhysFrame.containingAddress(self.getAddr());
+    }
+
+    /// Map the entry to the specified physical address
+    pub fn setAddr(self: *PageTableEntry, addr: PhysAddr) void {
+        //std.debug.assert(addr.isAligned(PageSize.Size4KiB.bytes()));
```

Comment:
Commented out for now, since it works in 64-bit, but not 32-bit. And we are compiling `paging_32.asm` for 32-bit.

```
$ make clean; make run
./src/zasm.zig:734:64: error: expected type 'usize', found 'u64'
        std.debug.assert(addr.isAligned(PageSize.Size4KiB.bytes()));
                                                               ^
./src/zasm.zig:734:64: note: unsigned 32-bit int cannot represent all possible unsigned 64-bit values
        std.debug.assert(addr.isAligned(PageSize.Size4KiB.bytes()));
                                                               ^
./src/zasm.zig:623:38: error: expected type 'usize', found 'u64'
        return std.mem.isAligned(self.value, alignment);
                                     ^
./src/zasm.zig:623:38: note: unsigned 32-bit int cannot represent all possible unsigned 64-bit values
        return std.mem.isAligned(self.value, alignment);
                                     ^
make: *** [Makefile:18: bin/paging_32_zig.o] Error 1
rm bin/boot_32_zig.o
```

Reply:
Added TODO to re-enable assert for `setAddr` in ab5cc598b056197530ebcc6b8b77d391bfd731f6..